database/maintenance.go

```diff
@@ -32,12 +39,32 @@ func MaintainThorough() (err error) {
 }
 
 // MaintainRecordStates runs record state lifecycle maintenance on all storages.
-func MaintainRecordStates() error {
+func MaintainRecordStates(ctx context.Context) error { //nolint:gocognit
+	// TODO: Put this in the storage interface to correctly maintain on all storages.
+	// Storages might check for deletion and expiry in the query interface and not return anything here.
+
+	// listen for ctx cancel
+	stop := abool.New()
+	doneCh := make(chan struct{}) // for goroutine cleanup
+	defer close(doneCh)
+	go func() {
+		select {
+		case <-ctx.Done():
+		case <-doneCh:
+		}
+		stop.Set()
+	}()
```

Comment:
I'd just use a function to check for the Done() channel. This way you can get rid of the `stop` abool, the `doneCh` and the dedicated goroutine.

```suggestion
isStopped := func() bool {
    select {
        case <- ctx.Done():
            return true
        default:
            return false
}
```

Reply:
Ok. I wasn't sure about how to best go about this. I'd love to have a flag on ctx ;)